src/Field.tsx

```diff
@@ -64,17 +64,32 @@ export interface FieldConfig {
   innerRef?: (instance: any) => void;
 }
 
-export type FieldAttributes<T> = GenericFieldHTMLAttributes & FieldConfig & T;
+export type FieldAttributes<T> = GenericFieldHTMLAttributes &
+  FieldConfig &
+  T & { name: string };
 
-export function useField<Val = any>(name: string, type?: string) {
+export function useField<Val = any>(
+  propsOrFieldName: string | FieldAttributes<Val>
+) {
   const formik = useFormikContext();
   if (__DEV__) {
     invariant(
       formik,
       'useField() / <Field /> must be used underneath a <Formik> component or withFormik() higher order component'
     );
   }
-  return formik.getFieldProps<Val>(name, type);
+
+  if (isObject(propsOrFieldName)) {
+    if (process.env.NODE_ENV !== 'production') {
+      invariant(
+        (propsOrFieldName as FieldAttributes<Val>).name,
```

Comment:
`isObject` should be defined to narrow the type automatically, like:

```
type NonObject = string | number | boolean | symbol | null | undefined;
const isObject = <T extends object>(maybeObject: NonObject | T): maybeObject is T => 
    typeof value === 'object'; // or whatever isObject's test really is, maybe check for array?
// then
if (isObject(propsOrFieldName)) { // propsOrFieldName is inferred as FieldAttributes<Val> }
```

Reply:
Can you open a new issue for this?

Reply:
Done, I'll resolve this conversation

#1599